### PR TITLE
Add user and group so we don't run as root

### PR DESCRIPTION
**Pull request by joewarner**
Although I know that best practice is NOT to run your docker images as `root` I hadn't seen a good example of how to setup another user before. However I was just reviewing https://github.com/camunda/docker-camunda-bpm-platform/blob/next/Dockerfile which is our template for our `Dockerfile` here and I saw that they have now added `camunda:camunda" user and group, so I thought it was time to make this change.

I have named the user and group `bp3user` and `bp3global` respectively. I don't know if these are the best names or not, but going with `bp3:bp3` didn't seem quite right.

The command looks find to me although I am not absolutely sure what the `-D` option to `adduser` is for. I've made the assumption that the parameter is a group reference and not a user reference (hence I am supplying `bp3global`).

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -16,5 +16,9 @@ RUN apk add --no-cache \
 ENTRYPOINT ["./docker-entrypoint.sh"]
 WORKDIR /worker
 
+RUN addgroup -g 1000 -S bp3global && \
+    adduser -u 1000 -S bp3user -G bp3global -h /worker -s /bin/bash -D bp3global
```
Comment (bschlosser) [inline Dockerfile:20]:
```suggestion
    adduser -u 1000 -S bp3user -G bp3global -h /worker -s /bin/bash -D bp3user
```
The last arg is supposed to be the _LOGIN_

**Comment by bschlosser** [comment]
What about using `bp3` as the group? I agree with you its nice to have different names so you can tell what they are, but it doesn't quite match linux standards (its common for them to be the same)

The `-D` is to use the defaults, `--defaults` is the long form, and might be preferable.

**Comment by joewarner** [comment]
> What about using `bp3` as the group? I agree with you its nice to have different names so you can tell what they are, but it doesn't quite match linux standards (its common for them to be the same)
> 
> The `-D` is to use the defaults, `--defaults` is the long form, and might be preferable.

I'm ok with using `bp3` as the group. I did start out with that and then couldn't really make my mind up! Also changed to using long form `--defaults` with `user` rather than `group`.

File: Dockerfile (hunk 1 above)
Comment (joewarner) [reply Dockerfile:20]: Suggested changes all agreed.

**Comment by bschlosser** [comment]
Looks good to me, but I diidn't test it.

**Comment by jphillips-bp3** [comment]
@bschlosser I now get the following error:

`adduser: unrecognized option: defaults`

```
jphillips@MacBook-Pro c7-rest-worker % docker build -t bp3global/c7-rest-worker:local .
[+] Building 1.1s (9/10)                                                                                                                                                                                                       docker:desktop-linux
 => [internal] load build definition from Dockerfile                                                                                                                                                                                           0.0s
 => => transferring dockerfile: 841B                                                                                                                                                                                                           0.0s
 => [internal] load metadata for docker.io/library/alpine:3.18                                                                                                                                                                                 1.0s
 => [auth] library/alpine:pull token for registry-1.docker.io                                                                                                                                                                                  0.0s
 => [internal] load .dockerignore                                                                                                                                                                                                              0.0s
 => => transferring context: 212B                                                                                                                                                                                                              0.0s
 => [internal] load build context                                                                                                                                                                                                              0.0s
 => => transferring context: 42.58kB                                                                                                                                                                                                           0.0s
 => [1/5] FROM docker.io/library/alpine:3.18@sha256:3ddf7bf1d408188f9849efbf4f902720ae08f5131bb39013518b918aa056d0de                                                                                                                           0.0s
 => CACHED [2/5] RUN apk add --no-cache         bash         curl         openjdk17-jre-headless                                                                                                                                               0.0s
 => CACHED [3/5] WORKDIR /worker                                                                                                                                                                                                               0.0s
 => ERROR [4/5] RUN addgroup -g 1000 -S bp3 &&     adduser -u 1000 -S bp3user -G bp3 -h /worker -s /bin/bash --defaults bp3user                                                                                                                0.1s
------
 > [4/5] RUN addgroup -g 1000 -S bp3 &&     adduser -u 1000 -S bp3user -G bp3 -h /worker -s /bin/bash --defaults bp3user:
0.087 adduser: unrecognized option: defaults
0.089 BusyBox v1.36.1 (2024-06-12 11:52:11 UTC) multi-call binary.
0.089 
0.089 Usage: adduser [OPTIONS] USER [GROUP]
0.089 
0.089 Create new user, or add USER to GROUP
0.089 
0.089   -h DIR          Home directory
0.089   -g GECOS        GECOS field
0.089   -s SHELL        Login shell
0.089   -G GRP          Group
0.089   -S              Create a system user
0.089   -D              Don't assign a password
0.089   -H              Don't create home directory
0.089   -u UID          User id
0.089   -k SKEL         Skeleton directory (/etc/skel)
------
Dockerfile:19
--------------------
  18 |     
  19 | >>> RUN addgroup -g 1000 -S bp3 && \
  20 | >>>     adduser -u 1000 -S bp3user -G bp3 -h /worker -s /bin/bash --defaults bp3user
  21 |     USER bp3user
--------------------
ERROR: failed to solve: process "/bin/sh -c addgroup -g 1000 -S bp3 &&     adduser -u 1000 -S bp3user -G bp3 -h /worker -s /bin/bash --defaults bp3user" did not complete successfully: exit code: 1
```

**Comment by joewarner** [comment]
Change —defaults to -D should do the trick I thinkSent from my iPhoneOn 21 Oct 2024, at 17:29, Justin Phillips ***@***.***> wrote:﻿
I have locally tried to build the image and get the following error:
[+] Building 0.6s (8/9)                                                                                                                                                                                                        docker:desktop-linux
 => [internal] load build definition from Dockerfile                                                                                                                                                                                           0.0s
 => => transferring dockerfile: 841B                                                                                                                                                                                                           0.0s
 => [internal] load metadata for docker.io/library/alpine:3.18                                                                                                                                                                                 0.5s
 => [internal] load .dockerignore                                                                                                                                                                                                              0.0s
 => => transferring context: 212B                                                                                                                                                                                                              0.0s
 => [1/5] FROM ***@***.***:3ddf7bf1d408188f9849efbf4f902720ae08f5131bb39013518b918aa056d0de                                                                                                                           0.0s
 => [internal] load build context                                                                                                                                                                                                              0.0s
 => => transferring context: 32.72kB                                                                                                                                                                                                           0.0s
 => CACHED [2/5] RUN apk add --no-cache         bash         curl         openjdk17-jre-headless                                                                                                                                               0.0s
 => CACHED [3/5] WORKDIR /worker                                                                                                                                                                                                               0.0s
 => ERROR [4/5] RUN addgroup -g 1000 -S bp3 &&     adduser -u 1000 -S bp3user -G bp3 -h /worker -s /bin/bash --defaults bp3user                                                                                                                0.1s
------                                                                                                                                                                                                                                              
 > [4/5] RUN addgroup -g 1000 -S bp3 &&     adduser -u 1000 -S bp3user -G bp3 -h /worker -s /bin/bash --defaults bp3user:
0.102 adduser: unrecognized option: defaults
0.104 BusyBox v1.36.1 (2024-06-12 11:52:11 UTC) multi-call binary.
0.104 
0.104 Usage: adduser [OPTIONS] USER [GROUP]
0.104 
0.104 Create new user, or add USER to GROUP
0.104 
0.104   -h DIR          Home directory
0.104   -g GECOS        GECOS field
0.104   -s SHELL        Login shell
0.104   -G GRP          Group
0.104   -S              Create a system user
0.104   -D              Don't assign a password
0.104   -H              Don't create home directory
0.104   -u UID          User id
0.104   -k SKEL         Skeleton directory (/etc/skel)
------
Dockerfile:19
--------------------
  18 |     
  19 | >>> RUN addgroup -g 1000 -S bp3 && \
  20 | >>>     adduser -u 1000 -S bp3user -G bp3 -h /worker -s /bin/bash --defaults bp3user
  21 |     USER bp3user
--------------------
ERROR: failed to solve: process "/bin/sh -c addgroup -g 1000 -S bp3 &&     adduser -u 1000 -S bp3user -G bp3 -h /worker -s /bin/bash --defaults bp3user" did not complete successfully: exit code: 1

View build details: docker-desktop://dashboard/build/desktop-linux/desktop-linux/vz6t6u8wv1ghyl3av2ww96und
***@***.*** c7-rest-worker % 


—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you authored the thread.Message ID: ***@***.***>

**Comment by jphillips-bp3** [comment]
That now builds, and I can run the container and process requests.

**Comment by jphillips-bp3** [comment]
@bschlosser given this works can we merge this in?

**Pull request merged**